### PR TITLE
Changes in Memcached and Redis classes (drivers of the Cache library)

### PR DESCRIPTION
**Pull request by masterklavi**
1. Changed the mechanism of detecting the serialized values  
    The old mechanism does unnecessary requests to Redis server, and it may have some mistakes in unserializing of value because it doesn't check the type of value that can be different. For example, in the asynchronous requests to Redis with different types of data. The best way is determine the type of value (serialized or not) at the get method.
2. Added the destructor (disconnect from memcache(d))  
    Closing the all connections to cache server is a good practice
3. UNIX socket for connection  
    Many of people use the unix socket for connecting to Memcached because it is the fastest method for data exchange between application and the memcached server.
4. Docblock corrections  
    Some comments don't contain the variable name

### CODE REVIEW

File: system/libraries/Cache/drivers/Cache_memcached.php [hunk 1]
```diff
@@ -60,11 +60,13 @@ class CI_Cache_memcached extends CI_Driver {
...
-	protected $_memcache_conf = array(
+	protected $_config = array(
 		'default' => array(
-			'host'		=> '127.0.0.1',
-			'port'		=> 11211,
-			'weight'	=> 1
+			'socket_type'	=> 'tcp',
+			'host'			=> '127.0.0.1',
+			'port'			=> 11211,
+			'socket'		=> '/var/run/memcached/memcached.sock',
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Why is this the default?

Also, if you'll be changing the *alignment*, use spaces for it.

Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: And on another thought, why do we need a separate 'socket' element instead of just using 'host'?
We also don't need 'socket_type' ... it's very easy to determine whether the 'host' value is a hostname or a path on the filesystem.

File: system/libraries/Cache/drivers/Cache_memcached.php [hunk 2]
```diff
@@ -104,24 +97,35 @@ public function __construct()
 		{
 			$this->_memcached = new Memcache();
 		}
-		else
-		{
-			log_message('error', 'Cache: Failed to create Memcache(d) object; extension not loaded?');
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Why is this removed?

File: system/libraries/Cache/drivers/Cache_memcached.php [hunk 3]
```diff
@@ -104,24 +97,35 @@ public function __construct()
...
+			}
+			
+			if ($this->_memcached instanceof Memcache)
 			{
-				// Third parameter is persistance and defaults to TRUE.
 				$this->_memcached->addServer(
 					$cache_server['hostname'],
 					$cache_server['port'],
-					TRUE,
+					TRUE,						// persistance
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: I don't quite like that you're just moving comments around in the first place - we don't do changes for the sake of change - but if you do, prepend with a single space, not multiple tabs.

File: system/libraries/Cache/drivers/Cache_memcached.php [hunk 4]
```diff
@@ -169,16 +173,14 @@ public function save($id, $data, $ttl = 60, $raw = FALSE)
 			$data = array($data, time(), $ttl);
 		}
 
-		if (get_class($this->_memcached) === 'Memcached')
+		if ($this->_memcached instanceof Memcache)
 		{
-			return $this->_memcached->set($id, $data, $ttl);
+			return $this->_memcached->set($id, $data, 0, $ttl);
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: You wouldn't need to do this if you didn't invert the condition sequence ...

File: system/libraries/Cache/drivers/Cache_memcached.php [hunk 5]
```diff
@@ -286,4 +288,28 @@ public function is_supported()
...
+	/**
+	 * Class destructor
+	 *
+	 * Closes the connection to Memcached if present.
+	 *
+	 * @return	void
+	 */
+	public function __destruct()
+	{
+		if ($this->_memcached)
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: We don't really need this ... `$this->_memcached instanceof Memcache` won't trigger any errors, though you will have to do an `elseif ($this->_memcached instanceof Memcached)`.

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 6]
```diff
@@ -55,11 +55,12 @@ class CI_Cache_redis extends CI_Driver
...
 	 */
 	protected static $_default_config = array(
-		'socket_type' => 'tcp',
-		'host' => '127.0.0.1',
-		'password' => NULL,
-		'port' => 6379,
-		'timeout' => 0
+		'socket_type'	=> 'tcp',
+		'host'			=> '127.0.0.1',
+		'socket'		=> '/var/run/redis/redis.sock',
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Why are you introducing this and changing all the alignments?

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 7]
```diff
@@ -138,18 +130,24 @@ public function __construct()
 	/**
 	 * Get cache
 	 *
-	 * @param	string	Cache ID
+	 * @param	string	$id	Cache ID
 	 * @return	mixed
 	 */
-	public function get($key)
+	public function get($id)
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Do you really need to do this?

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 8]
```diff
@@ -138,18 +130,24 @@ public function __construct()
...
 	{
-		$value = $this->_redis->get($key);
-
-		if ($value !== FALSE && isset($this->_serialized[$key]))
+		$value = $this->_redis->get($id);
+		
+		$unserialized = @unserialize($value);
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: No! We don't do blind `unserialize()` calls.

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 9]
```diff
@@ -138,18 +130,24 @@ public function __construct()
...
+		if ($unserialized !== FALSE)
 		{
-			return unserialize($value);
+				return $unserialized;
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Why change indentation?

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 10]
```diff
@@ -138,18 +130,24 @@ public function __construct()
...
 		}
 
+		if ($value === serialize(FALSE))
+		{
+				return FALSE;
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: This too is one level too far in indentation.

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 11]
```diff
@@ -166,21 +164,10 @@ public function get($key)
 	 */
 	public function save($id, $data, $ttl = 60, $raw = FALSE)
 	{
-		if (is_array($data) OR is_object($data))
+		if ( ! is_string($data))
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: What if it's an integer, float, boolean or null?

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 12]
```diff
@@ -190,23 +177,12 @@ public function save($id, $data, $ttl = 60, $raw = FALSE)
...
-		}
-
-		if (isset($this->_serialized[$key]))
-		{
-			$this->_serialized[$key] = NULL;
-			$this->_redis->sRemove('_ci_redis_serialized', $key);
-		}
-
-		return TRUE;
+		return $this->_redis->delete($id) === 1;
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: We always put parentheses around such conditions.

File: system/libraries/Cache/drivers/Cache_redis.php [hunk 13]
```diff
@@ -271,17 +247,17 @@ public function cache_info($type = NULL)
 	/**
 	 * Get cache metadata
 	 *
-	 * @param	string	Cache key
+	 * @param	string	$id	Cache ID
 	 * @return	array
 	 */
-	public function get_metadata($key)
+	public function get_metadata($id)
```
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Again a completely unnecessary change.

**Comment by narfbg** [comment]
Please don't submit such large changesets at once.

As it stands - I'd merge the Memcache(d) modifications after the patch is refactored to meed our standards, but I can't because you're submitting it together with Redis changes that I don't want.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 1 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: > And on another thought, why do we need a separate 'socket' element instead of just using 'host'?

Because it is more correct. UNIX socket is not host+port. The socket parameter is also specified in Redis (driver in cache library).

> it's very easy to determine whether the 'host' value is a hostname or a path on the filesystem.

I think it is a bad idea. Host can't be a _file path_, it can be a _hostname_ only.

> Also, if you'll be changing the alignment, use spaces for it.

Ok, I'll fix them.

> Why is this the default?

Unfortunately there is no exact default value for the socket so we can use an empty string.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 2 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Because we have the "is_supported()" method

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 3 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: > if you do, prepend with a single space, not multiple tabs.

Ok, I got it

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 4 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: I think it is not a good practice to use different conditions for similar things so I invert them.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 5 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: > We don't really need this ... $this->_memcached instanceof Memcache won't trigger any errors, 

I don't understand what errors was it trigger? In version 5.2 and more it won't trigger any errors.

> though you will have to do an elseif ($this->_memcached instanceof Memcached).

I think it doesn't need. The "_memcached" property contains object of Memcache or Memcached _only_.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 6 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: 1. There was no the socket parameter so `$success = $this->_redis->connect($config['socket']);` will trigger an error.
2. The default value for UNIX socket is `/var/run/redis/redis.sock` (it may be `/tmp/redis.sock` also).

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 7 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Yes. Why does it used different names of one entity?

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 8 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Look at the code again. I didn't do blind `unserialize()` calls. I check it efficiently. There are no any errors. Rather the old mechanism has some mistakes.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 11 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: If it is an integer, float, boolean or null we store data as serialized. Next if we'll get this data the `get()` method will return the correct value (not a string, because the string is not a correct value for integer, float, boolean and null).

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 12 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok, I got it

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 13 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Unnecessary? Why does it used different names of one entity?

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 1 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: >> And on another thought, why do we need a separate 'socket' element instead of just using 'host'?
>
> Because it is more correct. UNIX socket is not host+port. The socket parameter is also specified in Redis (driver in cache library).

There's no such thing as "more correct" in this situation ... we just don't need 2 variables if we could do just fine with only one. Especially given that the Memcache(d) classes too use the same parameter to accept either a hostname or a unix socket path.

>> it's very easy to determine whether the 'host' value is a hostname or a path on the filesystem.
>
> I think it is a bad idea. Host can't be a file path, it can be a hostname only.

Read above.

>> Why is this the default?
>
> Unfortunately there is no exact default value for the socket so we can use an empty string.

`NULL` instead of an empty string, actually. But that's not the point ... that the underlying methods don't have defaults doesn't mean that we should invent them.
Erroring because of a non-existing value is way better (read "user-friendlier") than erroring because of a default value that the user didn't want to use.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 2 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: `is_supported()` neither logs this condition, nor is it supposed to.
This log message isn't here by accident.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 4 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Your reply here doesn't make any sense ... the order inversion provides absolutely nothing in any way imaginable.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 5 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: >> We don't really need this ... $this->_memcached instanceof Memcache won't trigger any errors,
>
> I don't understand what errors was it trigger? In version 5.2 and more it won't trigger any errors.

That's what I'm saying - you're not avoiding any errors with this `if` condition and therefore it is unnecessary.

... and that gives context to the other part of my comment:

>> though you will have to do an elseif ($this->_memcached instanceof Memcached).
>
> I think it doesn't need. The $this->_memcached property contains object of Memcache or Memcached only.

You'll need it when `if ($this->_memcached)` is removed.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 6 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: OK, we should fix the non-existing 'socket' problem, but my opinion about socket path defaults is the same here as for Memcache(d).

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 1 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Well, then it is needed to _remove_ the socket parameter from Redis driver (cache library).

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 2 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Ok, I got It. I'll return `log_message()`.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 7 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Somebody wrote it that way. The name is completely irrelevant, wouldn't matter if it was called `$fubar`.

Sure, it may be inconsistent compared to other classes, but that doesn't make it incorrect, not does it make this change necessary.

If you submitted a separate pull request targeted *only* at cleaning up such inconsistencies, I'd probably merge it. But this patch is already quite large and hard to review enough, we don't need stylistic chances on top of it to make the diff even longer and harder to review.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 4 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: I don't deny this. But it will be great if we will use a nice code.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 5 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Ok

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 8 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: I'm looking at it and you *are* calling it blindly *at this point* right here and there's nothing efficient about it.
The check that follows it is a band-aid, not a solution - you need to *know* that the value was serialized before calling `unserialize()` on it.

Also, there are errors. You're just suppressing them - we don't do that unless there's absolutely no other way to make something work and that's not the case here.

If you think that the current mechanism has "some mistakes" - please explain them first.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 6 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok, we can remove the `socket` and `socket_type` parameters.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 11 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: That's the thing I don't like about it - we *don't* need to serialize scalar types, only arrays and objects.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 13 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Same as here: https://github.com/bcit-ci/CodeIgniter/pull/4515#discussion_r55832527

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 7 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok, I'll do it in the next pill request.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 1 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: While I agree that it should be removed from Redis, it's not necessarily "needed" and that should happen in a *separate* pull request.

Nothing will break if this driver is patched and the other one is not. You're just making the changes harder to reason about by mixing so much things at once.

You can't "fix" everything at once ... bug fixes, new features, stylistic changes don't belong together, even less so on separate pieces of code that are completely independent from each other.

Please refer to what I said here: https://github.com/bcit-ci/CodeIgniter/pull/4515#issuecomment-195288759

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 4 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_memcached.php]: How is checking for Memcache before Memcache**d** make this "nice"?

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 8 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: > The check that follows it is a band-aid, not a solution - you need to know that the value was serialized before calling unserialize() on it.

It isn't a band-aid and it isn't my idea. Look at http://php.net/manual/en/function.unserialize.php#85097

> If you think that the current mechanism has "some mistakes" - please explain them first.

I wrote them in my first message:

> The old mechanism does unnecessary requests to Redis server, and it may have some mistakes in unserializing of value because it doesn't check the type of value that can be different. For example, in the asynchronous requests to Redis with different types of data.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 11 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok. What would we _get_ if we _set_ a boolean value like `FALSE` or a `NULL` value? We will _get_ the same value `""`. 
Worth mentioning that `"" !== FALSE` and `"" !== NULL`.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 8 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: >> The check that follows it is a band-aid, not a solution - you need to know that the value was serialized before calling unserialize() on it.
>
> It isn't a band-aid and it isn't my idea. Look at http://php.net/manual/en/function.unserialize.php#85097

It is a band-aid. What I mean by "band-aid" is that it is masking a failure (to unserialize) that has already happened.

I don't undestand why someone writing a comment on php.net should make something a better idea, if that's what you're implying.
`serialize()` and `unserialize()` are not as harmless as you may think they are, you shouldn't be using them just to make quick checks (ironically, they are also not quick).

>> If you think that the current mechanism has "some mistakes" - please explain them first.
>
> I wrote them in my first message:
>

See, this is what wouldn't happen if this wasn't one huge patch - we're discussing so much things at once that I've completely lost your initial message. :)

>> The old mechanism does unnecessary requests to Redis server, and it may have some mistakes in unserializing of value because it doesn't check the type of value that can be different. For example, in the asynchronous requests to Redis with different types of data.

I disagree with the first part of this and don't really understand what you mean with the second part. Fair enough - you may be trying to solve a problem that I don't see - but please do it in a separate pull request where we can focus on it alone.

**Comment by masterklavi** [comment]
> Please don't submit such large changesets at once.
> As it stands - I'd merge the Memcache(d) modifications after the patch is refactored to meed our standards, but I can't because you're submitting it together with Redis changes that I don't want.

Ok

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 11 above)
Comment (narfbg) [inline system/libraries/Cache/drivers/Cache_redis.php]: Would we? I'm quite certain that Redis stores NULLs as NULLs and booleans as booleans.

Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Yeah, sorry, I'm confused with memcache.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 10 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 9 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok, I got it.

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 13 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: Ok

File: system/libraries/Cache/drivers/Cache_redis.php (hunk 8 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_redis.php]: > It is a band-aid. What I mean by "band-aid" is that it is masking a failure (to unserialize) that has already happened.

Ok, then I agree. I'll do another solution in a separate pull request.

> I don't undestand why someone writing a comment on php.net should make something a better idea, if that's what you're implying.

Ok.

> serialize() and unserialize() are not as harmless as you may think they are, you shouldn't be using them just to make quick checks (ironically, they are also not quick).

Yes, but it faster than an additional request to Redis server `$serialized = $this->_redis->sMembers('_ci_redis_serialized')`.

> See, this is what wouldn't happen if this wasn't one huge patch - we're discussing so much things at once that I've completely lost your initial message. :)

Yeah, I'm sorry. It is my first pull request.

> I disagree with the first part of this and don't really understand what you mean with the second part. Fair enough - you may be trying to solve a problem that I don't see - but please do it in a separate pull request where we can focus on it alone.

Ok.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 1 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: Ok.

File: system/libraries/Cache/drivers/Cache_memcached.php (hunk 4 above)
Comment (masterklavi) [inline system/libraries/Cache/drivers/Cache_memcached.php]: It is nice if we use `if ($this->_memcached instanceof Memcache) ... else ...` in `_construct()`, `save()`, and etc.
It isn't nice if we use `if ($this->_memcached instanceof Memcache) ... else ...` in `_construct()`, and  `if ($this->_memcached instanceof MemcacheD) ... else ...` in `save()`.
